### PR TITLE
feat: more unhandled effect exception information

### PR DESCRIPTION
**Pull request by JonathanStarup**

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/jvm/BackendObjType.scala [hunk 1]
```diff
@@ -203,7 +203,7 @@ object BackendObjType {
           // get expression as thunk
           DUP() ~ GETFIELD(ExpField) ~ CHECKCAST(Thunk.jvmName) ~
           // this.value = thunk.unwind()
-          Result.unwindSuspensionFreeThunkToType(tpe, SourceLocation.Unknown) ~ PUTFIELD(ValueField) ~
+          Result.unwindSuspensionFreeThunkToType(tpe, Some("in lazy expression"), SourceLocation.Unknown) ~ PUTFIELD(ValueField) ~
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/BackendObjType.scala:206]: during call to Lazy.force.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/BackendObjType.scala [hunk 2]
```diff
@@ -339,105 +339,105 @@ object BackendObjType {
...
-              Result.unwindSuspensionFreeThunkToType(JavaObject.toTpe, SourceLocation.Unknown) ~ ARETURN()
+              Result.unwindSuspensionFreeThunkToType(JavaObject.toTpe, Some("in Object Function interface method"), SourceLocation.Unknown) ~ ARETURN()
           ))
         case ObjConsumer => InstanceMethod(this.jvmName, IsPublic, IsFinal, "accept",
           mkDescriptor(JavaObject.toTpe)(VoidableType.Void),
           Some(_ =>
             thisLoad() ~
               DUP() ~ ALOAD(1) ~ PUTFIELD(ArgField(0)) ~
-              Result.unwindSuspensionFreeThunkToType(JavaObject.toTpe, SourceLocation.Unknown) ~ RETURN()
+              Result.unwindSuspensionFreeThunkToType(JavaObject.toTpe, Some("in Object Consumer interface method"), SourceLocation.Unknown) ~ RETURN()
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/BackendObjType.scala:349]: in java.util.function.Consumer.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/jvm/BackendObjType.scala:349]: i.e. let us use the actual full Java name.

File: main/src/ca/uwaterloo/flix/language/phase/jvm/BackendObjType.scala [hunk 3]
```diff
@@ -339,105 +339,105 @@ object BackendObjType {
...
-              Result.unwindSuspensionFreeThunkToType(BackendType.Bool, SourceLocation.Unknown) ~ IRETURN()
+              Result.unwindSuspensionFreeThunkToType(BackendType.Bool, Some("in Object Predicate interface method"), SourceLocation.Unknown) ~ IRETURN()
           ))
         case IntFunction => InstanceMethod(this.jvmName, IsPublic, IsFinal, "apply",
           mkDescriptor(BackendType.Int32)(JavaObject.toTpe),
           Some(_ =>
             thisLoad() ~
               DUP() ~ ILOAD(1) ~ PUTFIELD(ArgField(0)) ~
-              Result.unwindSuspensionFreeThunkToType(JavaObject.toTpe, SourceLocation.Unknown) ~ ARETURN()
+              Result.unwindSuspensionFreeThunkToType(JavaObject.toTpe, Some("in Int Function interface method"), SourceLocation.Unknown) ~ ARETURN()
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/BackendObjType.scala:363]: i.e. java.util.function.IntFunction

File: main/src/ca/uwaterloo/flix/language/phase/jvm/BackendObjType.scala [hunk 4]
```diff
@@ -1562,7 +1562,7 @@ object BackendObjType {
     def InvokeMethod: InterfaceMethod = InterfaceMethod(this.jvmName, "invoke", mkDescriptor()(Result.toTpe))
 
     def RunMethod: DefaultMethod = DefaultMethod(this.jvmName, IsPublic, NotFinal, "run", mkDescriptor()(VoidableType.Void), Some(_ =>
-      thisLoad() ~ Result.unwindSuspensionFreeThunk(SourceLocation.Unknown) ~ POP() ~ RETURN()
+      thisLoad() ~ Result.unwindSuspensionFreeThunk(Some("in Run interface method"), SourceLocation.Unknown) ~ POP() ~ RETURN()
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/BackendObjType.scala:1565]: Runnable?

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenAnonymousClasses.scala [hunk 5]
```diff
@@ -150,7 +150,7 @@ object GenAnonymousClasses {
       }
 
       // Invoke the closure
-      BackendObjType.Result.unwindSuspensionFreeThunkToType(BackendType.toErasedBackendType(tpe), loc)(new BytecodeInstructions.F(methodVisitor))
+      BackendObjType.Result.unwindSuspensionFreeThunkToType(BackendType.toErasedBackendType(tpe), Some("in anonymous class method"), loc)(new BytecodeInstructions.F(methodVisitor))
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenAnonymousClasses.scala:153]: We should know both class name and method name.

**Review by magnus-madsen — COMMENTED** [review]
I think this is a great idea. I see three overall improvements:

- Let us drop the Option and require a string.
- Let us use more precise language (see my comments).
- Use real source locations.

**Comment by JonathanStarup** [comment]
Regarding your comment about real source positions, places like `lazy.force` don't really have a source position since it's just a generally used generated class unless you want to materialize the source location for each use of a lazy value. 

The `SourceLocation.Unknown` used now are places like this where there is not a meaningful source location.

**Comment by magnus-madsen** [comment]
> Regarding your comment about real source positions, places like `lazy.force` don't really have a source position since it's just a generally used generated class unless you want to materialize the source location for each use of a lazy value.
> 
> The `SourceLocation.Unknown` used now are places like this where there is not a meaningful source location.

But why not? Surely the code comes from somewhere?

**Comment by JonathanStarup** [comment]
Lazy$Obj.force is used a bunch of places, there is no static source location

**Review by magnus-madsen — APPROVED** [review]

**Pull request merged**